### PR TITLE
Fix build fail due to hardcoded directory

### PR DESCRIPTION
**Pull request by ParzivalWolfram**
Pretty simple fix, might as well make it official.

### CODE REVIEW

**Review by nickdesaulniers — APPROVED** [review]

**Comment by nickdesaulniers** [comment]
yes, the rootfs.cpio.gz can be placed in the kernel directory prior to building.

**Comment by ParzivalWolfram** [comment]
> yes, the rootfs.cpio.gz can be placed in the kernel directory prior to building.

I can't seem to get that to work in buildroot. It still puts out a zImage with no rootfs.

**Comment by nickdesaulniers** [comment]
ok, let me see if I can reproduce. I'll try to build this repo tonight.

File: arch/arm/configs/nintendo3ds_defconfig [hunk 1]
```diff
@@ -147,7 +147,7 @@ CONFIG_SCHED_AUTOGROUP=y
 # CONFIG_SYSFS_DEPRECATED is not set
 CONFIG_RELAY=y
 CONFIG_BLK_DEV_INITRD=y
-CONFIG_INITRAMFS_SOURCE="/home/wolfvak/linux-3ds/rootfs"
+# CONFIG_INITRAMFS_SOURCE is not set
```
Comment (nickdesaulniers) [inline arch/arm/configs/nintendo3ds_defconfig:150]: ah, so when I go in via menuconfig and remove this, it gets set to:

CONFIG_INITRAMFS_SOURCE=""

but I see the same panic you do with that.

I think we want

```diff
-CONFIG_INITRAMFS_SOURCE="/home/wolfvak/linux-3ds/rootfs"
+CONFIG_INITRAMFS_SOURCE="rootfs.cpio.gz"
```

Then that boots for me (though it seems that the bottom screen isn't initialized, and is missing the keyboard!)

**Review by nickdesaulniers — CHANGES_REQUESTED** [review]

**Review by nickdesaulniers — APPROVED** [review]
Did this fix things for you? Were you able to boot?

**Comment by ParzivalWolfram** [comment]
> Did this fix things for you? Were you able to boot?

No, however I did spot the following in buildroot's output:
```
# As the kernel gets compiled before root filesystems are
# built, we create a fake cpio file. It'll be
# replaced later by the real cpio archive, and the kernel will be
# rebuilt using the linux-rebuild-with-initramfs target.
```
Manually invoking `make linux-rebuild-with-initramfs` did regenerate the initramfs, but the image is still exactly the same when started: no rootfs mounted. I'm gonna try a few other things before I mess with this again, as i've hit a couple other issues with this build of Buildroot.

**Comment by nickdesaulniers** [comment]
FWIW; [this](https://github.com/ClangBuiltLinux/boot-utils/blob/main/buildroot/arm.config) is precisely the buildroot config I was using to test.  I wrote the initial one a long time ago, and have been somewhat hands on on maintaining them recently.  Buildroot has been great and low touch for our purposes of CI; so I'm a little rusty on some of the details around buildroot.

I definitely think it would be nice to have our own repo containing buildroot configs and overlays, perhaps even binaries.  But it's critical to agree upon something that works for everyone (or rather, is able to be confirmed as bootable by multiple people).

**Comment by ParzivalWolfram** [comment]
I'll test that config after I get done making food. At current, the latest
buildroot snapshot is choking on the same var, but because it's trying to
`make` it directly, and has "no rule to make rootfs.cpio.gz."

**Comment by ParzivalWolfram** [comment]
> FWIW; [this](https://github.com/ClangBuiltLinux/boot-utils/blob/main/buildroot/arm.config) is precisely the buildroot config I was using to test. I wrote the initial one a long time ago, and have been somewhat hands on on maintaining them recently. Buildroot has been great and low touch for our purposes of CI; so I'm a little rusty on some of the details around buildroot.
> 
> I definitely think it would be nice to have our own repo containing buildroot configs and overlays, perhaps even binaries. But it's critical to agree upon something that works for everyone (or rather, is able to be confirmed as bootable by multiple people).

Mine looks very similar, just... a lot longer.

**Comment by nickdesaulniers** [comment]
link?

**Comment by ParzivalWolfram** [comment]
Sorry for the late reply, didn't see this comment until now. Link to what?

**Comment by nickdesaulniers** [comment]
Your config

**Comment by ParzivalWolfram** [comment]
[config.zip](https://github.com/linux-3ds/linux/files/6250760/config.zip)

**Comment by nickdesaulniers** [comment]
> config.zip

It might be easier to see what's different/unique here if you use

```sh
$ cd buildroot
$ make savedefconfig
$ make -j$(nproc) savedefconfig BR2_DEFCONFIG=defconfig
```
then upload `defconfig`.  You can then rebuild via:
```sh
$ make -j$(nproc) BR2_DEFCONFIG=defconfig
```

Contrast that with:
```sh
#!/usr/bin/env sh
set -ex

BUILDROOT_VERSION=buildroot-2021.02

rm -f ${BUILDROOT_VERSION}.tar.gz
rm -rf ${BUILDROOT_VERSION}
rm -f rootfs.cpio.gz

# Download buildroot
wget https://buildroot.org/downloads/${BUILDROOT_VERSION}.tar.gz

# Build buildroot
tar xvf ${BUILDROOT_VERSION}.tar.gz
cd ${BUILDROOT_VERSION}/
BR2_DEFCONFIG=../arm.config make -j$(nproc) defconfig
make -j$(nproc)
cd ..

# Compress image
gzip ${BUILDROOT_VERSION}/output/images/rootfs.cpio
mv ${BUILDROOT_VERSION}/output/images/rootfs.cpio.gz .
# put rootfs.cpio.gz in your top level kernel sources and build zImage.
```
```
# arm.config
BR2_arm=y
BR2_TOOLCHAIN_EXTERNAL=y
BR2_TOOLCHAIN_EXTERNAL_BOOTLIN=y
BR2_TOOLCHAIN_EXTERNAL_BOOTLIN_ARMV5_EABI_UCLIBC_STABLE=y
BR2_TARGET_GENERIC_ROOT_PASSWD="root"
BR2_TARGET_GENERIC_GETTY_PORT="ttyAMA0"
BR2_TARGET_ROOTFS_CPIO=y
```
(Though some of your V6 VFPV2 additions might be nice to add).

Your config still has `BR2_KERNEL_HEADERS_CUSTOM_REPO_URL=` pointed to your kernel, so there's a chicken and the egg problem.  Your config says: "go build the kernel from my github fork."  But then your kernel says "build my initramfs from rootfs.cpio.gz" which is what the final build of buildroot would look like, which can't be built until the kernel is built.  It's a cyclic dependency.

I think BR2_TARGET_ROOTFS_INITRAMFS might be what you want rather than BR2_TARGET_ROOTFS_CPIO_GZIP=y, though that looks kind of hacky.

During bringup I think it makes more sense to use the kernel build system to drive zImage creation, rather than Buildroot.  Once drivers start landing in upstream Linux, then it makes more sense (or becomes much easier) to move to having buildroot build everything.

Having a solid config fragment file (see my directions above for generating it) is a good starting point that we can both take advantage of with these different approaches.  If you want to generate your fragment, we should collaborate on that in a new repo.

**Comment by nickdesaulniers** [comment]
I'm merging this for now so that we can enable separate DTB files for CTR and KTR in the booloader (#14).  You'll need to adjust that in your buildroot configs (there can now be one of 2 .dtb files named either nintendo3ds_ctr.dtb or nintendo3ds_ktr.dtb rather than one named ctr.dtb).  You'll want to use the KTR on New 3ds to get double RAM and CPU cores (once we have those working, see #1 and #2).

I'll update the kernel build instructions on [the wiki](https://github.com/linux-3ds/linux/wiki) for something simple that works for now, but we should collaborate on a more fully featured buildroot based solution, perhaps in a new repo.

**Comment by nickdesaulniers** [comment]
updated: https://github.com/linux-3ds/linux/wiki

**Pull request merged**